File: Data Structures & Algorithms/longest-increasing-path-in-matrix/submission-5.py
class Solution:
    def longestIncreasingPath(self, matrix: List[List[int]]) -> int:
        rows = len(matrix)
        cols = len(matrix[0])

        dirs = [[-1,0], [1, 0], [0, -1], [0, 1]]
        memo = {}

        def dfs(i, j, pre):
            if i < 0 or j < 0 or i == rows or j == cols or pre >= matrix[i][j]:
                return 0

            if (i,j) in memo:
                return memo[(i,j)]

            res = 1
            res = max(res, 1+ dfs(i+1, j, matrix[i][j]))
            res = max(res, 1+ dfs(i-1, j, matrix[i][j]))
            res = max(res, 1+ dfs(i, j+1, matrix[i][j]))
            res = max(res, 1+ dfs(i, j-1, matrix[i][j]))

            memo[(i, j)] = res
            return res

        LIS = 0
        for i in range(rows):
            for j in range(cols):
                LIS = max(LIS, dfs(i, j, -1))

        return LIS

File: Data Structures & Algorithms/max-area-of-island/submission-5.java
class Solution {

    int[][] dirs = {{-1,0}, {1, 0}, {0, -1}, {0, 1}};

    public int maxAreaOfIsland(int[][] grid) {
        int rows = grid.length, cols = grid[0].length;

        int max = 0;
        for(int r=0; r<rows; r++){
            for(int c = 0; c<cols; c++){
                if(grid[r][c] == 1){
                    max = Math.max(max, dfs(grid, r, c));
                }
            }
        }

        return max;
    }

    private int dfs(int[][] grid, int r, int c){
        if(r < 0 || c < 0 || r >= grid.length || c >= grid[0].length || grid[r][c] == 0){
            return 0;
        }

        grid[r][c] = 0;
        int res = 1;
        for(int[] dir:dirs){
            res += dfs(grid, r + dir[0], c + dir[1]);
        }

        return res;
    }
}
